File: rubygems/pray-cli/lib/pray/resolve_local.rb
# frozen_string_literal: true

module Pray
  module Resolve
    module_function

    def missing_local_embed_guidance(path)
      "Prayfile lists `local \"#{path}\"` but the file does not exist. " \
        "Create the file or remove the entry from Prayfile, then run `pray install`."
    end

    def resolve_local_file(project_root, declaration)
      path = File.join(project_root, declaration.path)
      unless File.exist?(path)
        if declaration.optional
          return ResolvedLocalFile.new(
            path: path,
            manifest_path: declaration.path,
            content: "",
            source_checksum: Hashing.sha256_prefixed(""),
            position: declaration.position,
            optional: true
          )
        end
        raise Error.resolution(missing_local_embed_guidance(declaration.path))
      end

      content = Hashing.normalize_line_endings(File.read(path))
      ResolvedLocalFile.new(
        path: path,
        manifest_path: declaration.path,
        content: content,
        source_checksum: Hashing.sha256_prefixed(content),
        position: declaration.position,
        optional: declaration.optional
      )
    end
  end
end

File: crates/pray-core/src/resolve_local.rs
use super::ResolvedLocalFile;
use crate::hashing::sha256_prefixed;
use crate::manifest::ManifestLocal;
use crate::resolve_exports::read_text;
use crate::{PrayError, PrayResult};
use std::path::Path;

pub fn missing_local_embed_guidance(path: impl AsRef<str>) -> String {
    let path = path.as_ref();
    format!(
        "Prayfile lists `local \"{path}\"` but the file does not exist. \
         Create the file or remove the entry from Prayfile, then run `pray install`."
    )
}

pub(crate) fn resolve_local_file(
    project_root: &Path,
    declaration: &ManifestLocal,
) -> PrayResult<ResolvedLocalFile> {
    let path = project_root.join(&declaration.path);
    if !path.exists() {
        if declaration.optional {
            return Ok(ResolvedLocalFile {
                path,
                manifest_path: declaration.path.clone(),
                content: String::new(),
                source_checksum: sha256_prefixed(b""),
                position: declaration.position.clone(),
                optional: true,
            });
        }
        return Err(PrayError::Resolution(missing_local_embed_guidance(
            &declaration.path,
        )));
    }
    let content = read_text(&path)?;
    Ok(ResolvedLocalFile {
        source_checksum: sha256_prefixed(content.as_bytes()),
        content,
        path,
        manifest_path: declaration.path.clone(),
        position: declaration.position.clone(),
        optional: declaration.optional,
    })
}
